src/ShallowWaterRosenbrock.jl

```diff
@@ -0,0 +1,252 @@
+  # 1.2: the bernoulli function
+  compute_bernoulli_potential!(ϕ,dΩ,Q,L2MMchol,u₁⋅u₁,h₁,g)
+  # 1.3: the potential vorticity
+  compute_potential_vorticity!(q₁,H1h,H1hchol,dΩ,R,S,h₁,u₁,f,n)
+  # 1.4: assemble the velocity residual
+  bᵤ₁(v) = ∫(-1.0*(q₁ - τ*u₁⋅∇(q₁))*(v⋅⟂(F,n)))dΩ + ∫(DIV(v)*ϕ)dω
+  # 1.5: assemble the depth residual
+  bₕ₁(q) = ∫(-q*DIV(F))dω
+
+  bₕᵤ₁((v,q)) = bᵤ₁(v) + bₕ₁(q)
+  Gridap.FESpaces.assemble_vector!(bₕᵤ₁, get_free_dof_values(duh₁), Y)
+
+  # Solve for du₁, dh₁ over a MultiFieldFESpace
+  ldiv!(Blfchol, get_free_dof_values(duh₁))
+
+  # update
+  get_free_dof_values(u₂) .= get_free_dof_values(uₘ) .+ dt₁ .* get_free_dof_values(du₁)
+  get_free_dof_values(h₂) .= get_free_dof_values(hₘ) .+ dt₁ .* get_free_dof_values(dh₁)
+
+  # 2.1: the mass flux
+  compute_mass_flux!(F,dΩ,V,RTMMchol,u₁*(2.0*h₁ + h₂)/6.0+u₂*(h₁ + 2.0*h₂)/6.0)
+  # 2.2: the bernoulli function
+  compute_bernoulli_potential!(ϕ,dΩ,Q,L2MMchol,(u₁⋅u₁ + u₁⋅u₂ + u₂⋅u₂)/3.0,0.5*(h₁ + h₂),g)
+  # 2.3: the potential vorticity
+  compute_potential_vorticity!(q₂,H1h,H1hchol,dΩ,R,S,h₂,u₂,f,n)
+  # 2.4: assemble the velocity residual
+  bᵤ₂(v) = ∫(-0.5*(q₁ - τ*u₁⋅∇(q₁) + q₂ - τ*u₂⋅∇(q₂))*(v⋅⟂(F,n)))dΩ + ∫(DIV(v)*ϕ)dω
+  # 2.5: assemble the depth residual
+  bₕ₂(q) = ∫(-q*DIV(F))dω
+
+  # add A*[du₁,dh₁] to the [du₂,dh₂] vector
+  bₕᵤ₂((v,q)) = bᵤ₂(v) + bₕ₂(q)
+  Gridap.FESpaces.assemble_vector!(bₕᵤ₂, get_free_dof_values(duh₂), Y)
+
+  mul!(y_wrk, Amat, get_free_dof_values(duh₁))
+  get_free_dof_values(duh₂) .= get_free_dof_values(duh₂) .- y_wrk
+
+  # solve for du₂, dh₂
+  ldiv!(Bchol, get_free_dof_values(duh₂))
+
+  # update yⁿ⁺¹
+  du₂, dh₂ = duh₂
+  get_free_dof_values(u₂) .= get_free_dof_values(u₁) .+ dt .* get_free_dof_values(du₂)
+  get_free_dof_values(h₂) .= get_free_dof_values(h₁) .+ dt .* get_free_dof_values(dh₂)
+end
+
+function new_vtk_step(Ω,file,hn,un,wn)
+  createvtk(Ω,
+            file,
+            cellfields=["hn"=>hn, "un"=>un, "wn"=>wn],
+            nsubcells=4)
+end
+
+function compute_mean_depth!(wrk, L2MM, h)
+  # compute the mean depth over the sphere, for use in the approximate Jacobian
+  mul!(wrk, L2MM, get_free_dof_values(h))
+  h_int = sum(wrk)
+  wrk  .= 1.0
+  tmp   = L2MM*wrk # create a new vector, only doing this once during initialisation
+  a_int = sum(tmp)
+  h_avg = h_int/a_int
+  println("mean depth: ", h_avg)
+  h_avg
+end
+
+function shallow_water_rosenbrock_time_stepper(model, order, degree,
+                        h₀, u₀, f₀, g, H₀,
+                        λ, dt, τ, N; leap_frog=false,
+                        write_diagnostics=true,
+                        write_diagnostics_freq=1,
+                        dump_diagnostics_on_screen=true,
+                        write_solution=false,
+                        write_solution_freq=N/10,
+                        output_dir="nswe_eq_ncells_$(num_cells(model))_order_$(order)_rosenbrock")
+
+  # Forward integration of the shallow water equations
```

Comment:
Might we factor out in functions some common operations among both solvers performed right before the time integration loop?

Reply:
Have moved the function space initialisation and mass matrix assembly and LU factorisation into separate functions in Helpers, and using these in both solvers